src/validation.cpp

```diff
@@ -2772,11 +2772,14 @@ bool ResetBlockFailureFlags(CBlockIndex *pindex) {
 
 CBlockIndex* CChainState::AddToBlockIndex(const CBlockHeader& block)
 {
+    AssertLockHeld(cs_main);
+
     // Check for duplicate
     uint256 hash = block.GetHash();
-    BlockMap::iterator it = mapBlockIndex.find(hash);
-    if (it != mapBlockIndex.end())
-        return it->second;
+    CBlockIndex* pindex = LookupBlockIndex(hash);
```

Comment:
Again, please do not use LookupBlockIndex here (and a few other places). If the next step is to make LookupBlockIndex return a const CBlockIndex* (which I think it clearly is), this part of the patch just has to get reverted. There is no reason to avoid using mapBlockIndex in CChainState members, that's what CChainState is there for :).

Reply:
> There is no reason to avoid using mapBlockIndex in CChainState members, that's what CChainState is there for

At that point there can be `CBlockIndex* CChainState::LookupBlockIndex()`?

Reply:
Why? I see no reason to have a LookupBlockIndex at all *except* to avoid exposing mapBlockIndex/non-const CBlockIndexes outside of CChainState/validation.cpp. If we're inside of CChainState, doing a LookupBlockIndex only serves to add needless indirection.